Units Page for Displaying, Creating and Editing Units

andrew-cline opened the pull request:
# Description
Adds a new page for displaying all units currently stored in the database. This page allows users to create new units and edit new units via popup modal UI elements.

## Type of change

- [ ] Note merging this changes the node modules
- [ ] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

- Need to route adding a unit through the state. Otherwise, you need to refresh the page for new units to appear.
- When changing the type of unit, the edit functionality will fail, producing this error from the database 
 `oed-3-database-1  | 2022-04-27 02:23:01.337 UTC [59] ERROR:  duplicate key value violates unique constraint "units_type_of_unit_unit_index_key"
oed-3-database-1  | 2022-04-27 02:23:01.337 UTC [59] DETAIL:  Key (type_of_unit, unit_index)=(meter, 4) already exists.
oed-3-database-1  | 2022-04-27 02:23:01.337 UTC [59] STATEMENT:  UPDATE units SET name = 'Unit Name', identifier = 'Identifier', unit_represent = 'quantity', sec_in_rate = 3600, type_of_unit = 'meter', unit_index = 4, suffix = 'Suffix', displayable = 'none', preferred_display = false, note = 'Note' WHERE id = 22;`

Commit:
added files for admin unit page

andrew-cline committed:
Merge branch 'OpenEnergyDashboard:resourceGeneralization' into resourceGeneralization

Commit:
added front-end starter code for admin units page

Commit:
added necessary files for displaying units

Commit:
finished unit display page, added page and form for adding units

Commit:
Merge branch 'resourceGeneralization' of https://github.com/OpenEnergyDashboard/OED into resourceGeneralization

JustLe56 committed:
added ES and FR translation messages for unit page

Commit:
edited UnitsDetailContainer

Commit:
edited UnitsDetailContainer

sagarprasad63574 committed:
edited units details component, container, etc

sagarprasad63574 committed:
Merge pull request #1 from andrew-cline/sagar

edited units details component, container, etc

JustLe56 committed:
added header and footer to UnitDetailComponent

andrew-cline committed:
Merge pull request #2 from andrew-cline/resourceGenJustin

added header and footer to UnitDetailsComponent

sagarprasad63574 committed:
testing

sagarprasad63574 committed:
Merge branch 'resourceGeneralization' of https://github.com/andrew-cline/OED into sagar

sagarprasad63574 committed:
IDK

sagarprasad63574 committed:
Merge pull request #3 from andrew-cline/sagar

Sagar

sagarprasad63574 committed:
Added the button to identifier. Need to fix the logic

JustLe56 committed:
added SecInRate update; updates in client but does not save to database yet

Commit:
added button to unitRepresent column in UnitViewComponent

andrew-cline committed:
Merge pull request #4 from andrew-cline/sagar

Added the button to identifier. Need to fix the logic

andrew-cline committed:
Merge branch 'resourceGeneralization' into resourceGenJustin

andrew-cline committed:
Merge pull request #5 from andrew-cline/resourceGenJustin

Add SecInRate edit button

andrew-cline committed:
Revert "Add SecInRate edit button"

andrew-cline committed:
Merge pull request #6 from andrew-cline/revert-5-resourceGenJustin

Revert "Add SecInRate edit button"

andrew-cline committed:
Revert "Added the button to identifier. Need to fix the logic"

andrew-cline committed:
Merge pull request #7 from andrew-cline/revert-4-sagar

Revert "Added the button to identifier. Need to fix the logic"

Commit:
Merge branch 'resourceGeneralization' of https://github.com/andrew-cline/OED into Andrew

Hardik-Kharpude committed:
added a  button to the notes column in units page

andrew-cline committed:
Merge pull request #8 from andrew-cline/Andrew

Andrew

Commit:
merged resourceGenJustin with Andrew

andrew-cline committed:
Merge pull request #9 from andrew-cline/Andrew

merged resourceGenJustin with Andrew

Commit:
added UnitRepresent button

andrew-cline committed:
Merge pull request #10 from andrew-cline/Andrew

added UnitRepresent button

sagarprasad63574 committed:
Made changes to the backend

sagarprasad63574 committed:
Edited the backend and updated the code to work with unit represent and sec in rate

andrew-cline committed:
Merge branch 'resourceGeneralization' into HardikNoteButton

andrew-cline committed:
Merge pull request #12 from andrew-cline/HardikNoteButton

added a  button to the notes column in units page

Commit:
fixed somethings leftover from merge

Commit:
edited routes to include check for admin privledge

andrew-cline committed:
Merge pull request #13 from andrew-cline/Andrew

edited routes to include check for admin privledge

Commit:
added validation for secInRate

andrew-cline committed:
Merge pull request #14 from andrew-cline/Andrew

added validation for secInRate

JustLe56 committed:
fixed status message displaying on all columns

JustLe56 committed:
fixed checkbox; add submitNewUnit; changed unit id and identifer definition, added addUnit func to api

Commit:
fixed bug in unit create form

Commit:
fixed spelling in translation file

JustLe56 committed:
misc text fixes

JustLe56 committed:
added eng message for create unit container

JustLe56 committed:
fixes default selected dropdowns not triggering onchange

andrew-cline committed:
Merge branch 'OpenEnergyDashboard:resourceGeneralization' into resourceGeneralization

andrew-cline committed:
Merge branch 'resourceGeneralization' into sagar

andrew-cline committed:
Merge pull request #11 from andrew-cline/sagar

Made changes to the backend

Commit:
Merge branch 'justin_add_units' of https://github.com/andrew-cline/OED into resourceGeneralization

Commit:
Merge branch 'resourceGeneralization' of https://github.com/andrew-cline/OED into resourceGeneralization

Commit:
added formatted messages to createUnitComponent

Commit:
fixed preferredDisplay in UnitViewComponent

Commit:
Merge branch 'resourceGeneralization' of https://github.com/OpenEnergyDashboard/OED into resourceGeneralization

Commit:
removed duplicate unit files

Commit:
updated app.js to use units.js

Commit:
added jsonSchema import to units.js

JustLe56 committed:
changed add unit link to a modal display; need to add form content

JustLe56 committed:
adds modal create unit form; currently not working needs container

JustLe56 committed:
fixed imports

JustLe56 committed:
modal variation using hooks, does not actually modify database yet

JustLe56 committed:
create unit modal successfully submits to database; updated create unit form defaults

sagarprasad63574 committed:
UnitModelEdit

JustLe56 committed:
polished code formatting and added some comments

JustLe56 committed:
fixed unit state issue

Commit:
fixed unit state issue

JustLe56 committed:
Merge branch 'justin_createUnitModal' of https://github.com/andrew-cline/OED into clean_oed_resgen

JustLe56 committed:
Merge branch 'Unit_Dropdown' of https://github.com/andrew-cline/OED into createUnitModalCopy

sagarprasad63574 committed:
Merge branch 'resourceGeneralization' of https://github.com/andrew-cline/OED into UnitModelEdit

Commit:
edited submitNewUnit and addUnit route

sagarprasad63574 committed:
Working with unit model on the unit page and unit edit. Some problems are in order to update the unit in the database, need to click the save changes button twice(not sure where the problem is occurring)

JustLe56 committed:
lint changes

JustLe56 committed:
more linter changes

JustLe56 committed:
Merge branch 'resourceGeneralization' of https://github.com/OpenEnergyDashboard/OED into createUnitModal_UnitDropdown_merged

JustLe56 committed:
included edit note in render

JustLe56 committed:
added comments and lint changes

sagarprasad63574 committed:
Bug Fixes and Editing the modal for the units page and edit unit

sagarprasad63574 committed:
Bug Fixes

JustLe56 committed:
Merge branch 'UnitModelEdit' of https://github.com/andrew-cline/OED into modalEdit_modalCreate

JustLe56 committed:
Merge branch 'UnitModelEdit' of https://github.com/andrew-cline/OED into modalEdit_modalCreate

JustLe56 committed:
added linter fixes

JustLe56 committed:
added documentation and cleaned up comments

sagarprasad63574 committed:
Bug fixes

sagarprasad63574 committed:
Merge branch 'modalEdit_modalCreate' of https://github.com/andrew-cline/OED into modalEdit_modalCreate

JustLe56 committed:
linter fixes

huss commented on .vscode/settings.json:
```diff
@@ -0,0 +1,2 @@
+{
```
I don't think this edit should be part of the PR. I'm removing it.

huss commented on src/client/app/actions/units.ts:
```diff
@@ -35,3 +55,39 @@ export function fetchUnitsDetailsIfNeeded(): Thunk {
 	};
 }
+
+export function submitEditedUnits(): Thunk {
```
Can two units be edited at once? I'm just wondering if this is needed.

sagarprasad63574 replied:
No you cannot edit more than one unit at a time but the function submitEditedUnits() is needed to edit a unit. It would be best to change this function to edit a single unit.

huss commented on src/client/app/actions/units.ts:
```diff
@@ -35,3 +55,39 @@ export function fetchUnitsDetailsIfNeeded(): Thunk {
+/**
+ * Remove all the unit in editing without submitting them
+ */
+export function confirmEditedUnits(): Thunk {
```
Is this used anywhere? I see use of confirmEditedUnit.

sagarprasad63574 replied:
No. It can be removed

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -0,0 +1,160 @@
+import {FormattedMessage } from 'react-intl';
+import '../../styles/unit-add-modal.css';
+
+interface CreateUnitFormProps{
```
When I look at other props in OED, they do not put Create in front of the name. I would remove it.

andrew-cline replied:
Fixed in most recent commit

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -0,0 +1,160 @@
+interface CreateUnitFormProps{
+	name: string,
+	identifier: string,
+	unitRepresent: string,
```
unitRepresent, typeOfUnit & displayable are all enums in the DB but string here. Other places in the code it uses enum and not string. If you look at src/client/app/components/RouteComponent.tsx you see "defaultLanguage: LanguageTypes" which is defined in src/client/app/types/redux/i18n.ts. I think it would be better to use the enum rather than string to define these. The enums are defined in src/client/app/types/redux/units.ts. Also, that file has UnitData and UnitEditData (from you I think) that are very similar except for the types of these. Is the string version necessary?
Note: I just looked at the Edit modal and it does use the enum so it is a model for this one.

andrew-cline replied:
Fixed with most recent commit.

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -0,0 +1,160 @@
+					<div id="container">
+						<div id="modalChild">
+							{/* Modal content */}
+							<div className="containter-fluid">
```
Should this be container-fluid (spelling issue)?

andrew-cline replied:
Fixed in most recent commit

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -0,0 +1,160 @@
+										<div style={formInputStyle}>
+											<label><FormattedMessage id="unit.represent"/></label><br />
+											<Input type='select' onChange={({target}) =>  props.handleUnitRepresentChange(target.value)} required value={ props.unitRepresent}>
+												<option value='raw' key='raw'>raw</option>
```
It would be nice to get these values from the enum (see other comment for location). Same for type_of_unit and displayable.

andrew-cline replied:
Fixed with most recent commit.

huss commented on src/client/app/components/unit/UnitModalEditComponent.tsx:
```diff
@@ -0,0 +1,277 @@
+	 * (3) Call the function onHide() -> this will hide the modal and call function handleClose on UnitViewComponent.tsx */
+
+	private onSaveChanges() {
+		const oldIdentifer = this.props.unit.identifier;
```
oldIdentifer -> oldIdentifier.
oldunitRepresentType -> oldUnitRepresentType

andrew-cline replied:
Fixed in most recent commit

huss commented on src/client/app/components/unit/UnitModalEditComponent.tsx:
```diff
@@ -0,0 +1,277 @@
+	/**
+	 * The following handlers will change the state to the corresponding unit
+	 */
+	private handleIdentifier(event: React.ChangeEvent<HTMLInputElement>) {
```
Should this be handleIdentifierChange?

andrew-cline replied:
Fixed in most recent commit

huss commented on src/client/app/components/unit/UnitModalEditComponent.tsx:
```diff
@@ -0,0 +1,277 @@
+				<label>Type of Unit: </label>
+				<Input type='select' defaultValue={typeOfUnit}
+					onChange={event => this.handleTypeOfUnitChange(event)}>
+					<option value="unit"> unit </option>
```
All enum types should try to make menu based on enum. See src/client/app/components/admin/CreateUserComponent.tsx for an example.

andrew-cline replied:
Fixed with most recent commit.

huss commented on src/client/app/components/unit/UnitViewComponent.tsx:
```diff
@@ -0,0 +1,135 @@
+				</div>
+			)
+		}
+		return null;
```
Could you put in a comment on why this returns null?

andrew-cline replied:
Added comment for return statement in most recent commit.

huss commented on src/client/app/components/unit/UnitsDetailComponent.tsx:
```diff
@@ -0,0 +1,83 @@
+		const tooltipStyle = {
+			display: 'inline-block',
+			fontSize: '50%',
+			tooltipUnitView: loggedInAsAdmin? 'help.admin.unitview' : 'help.units.unitview'
```
This is a note to remind me to add the help info.

huss commented on src/client/app/components/RouteComponent.tsx:
```diff
@@ -33,6 +33,8 @@ import MapCalibrationContainer from '../containers/maps/MapCalibrationContainer'
 import { UserRole } from '../types/items';
 import { hasPermissions } from '../utils/hasPermissions';
+import UnitsDetailContainer from '../containers/unit/UnitsDetailContainer';
+import CreateUnitContainter from '../containers/unit/CreateUnitContainer';
```
CreateUnitContainter -> CreateUnitContainer

andrew-cline replied:
Fixed in most recent commit

huss commented on src/client/app/translations/data.js:
```diff
@@ -235,6 +236,29 @@ const localeData = {
 		"total": "total",
 		"undefined": "undefined",
+		"units": "Units",
+		"unit" : "Unit",
```
The way they are now alphabetized here needs to be copied to all the langauges.

OED does not use _ in these. They all need to be replaced with . (and in the code).

My quick scan indicates not all the strings are in all the languages.

JustLe56 replied:
Should be fixed as of 9d59efc.

huss commented on src/client/app/styles/unit-card-page.css:
```diff
@@ -0,0 +1,103 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+ .card-container {
+    /* border: 1px solid black; */
```
Should these commented out lines be here?

andrew-cline replied:
Removed those comments in most recent commit.

huss commented on src/client/app/components/unit/UnitViewComponent.tsx:
```diff
@@ -0,0 +1,135 @@
+					{this.props.unit.name}
+				</div>
+				<div className="unit-container">
+					<b>Unit Identifier:</b> {this.props.unit.identifier}
```
These strings should use the translated keys and translate.

JustLe56 replied:
Should be fixed as of 9d59efc.

huss commented on src/client/app/components/unit/UnitModalEditComponent.tsx:
```diff
@@ -0,0 +1,277 @@
+			<>
+				<Modal show={this.props.show} onHide={this.props.onhide}>
+					<Modal.Header closeButton>
+						<Modal.Title> Edit Unit Information</Modal.Title>
```
This should be translated.

JustLe56 replied:
Should be fixed as of 9d59efc.

huss commented on src/client/app/components/unit/UnitModalEditComponent.tsx:
```diff
@@ -0,0 +1,277 @@
+	private isIdentifier(identifier: string) {
+		return (
+			<div>
+				Identifier: <span><br /><input type="text" defaultValue={identifier}
```
All of these labels should be translated.

JustLe56 replied:
Should be fixed as of 9d59efc.

huss commented on src/client/app/translations/data.js:
```diff
@@ -33,6 +33,7 @@ const localeData = {
 		"create.group": "Create a Group",
 		"create.map": "Create a Map",
 		"create.user": "Create a User",
+		"create.unit": "Create a Unit",
```
You also have unit.create_new_unit which is used in the code but not in all languages. Pick one.

JustLe56 replied:
Should be fixed as of 8b4b241.

huss commented on src/client/app/components/unit/UnitModalEditComponent.tsx:
```diff
@@ -0,0 +1,277 @@
+	private isSecInRate(secInRate: number) {
+		return (
+			<div>
+				Sec In Rate: <span><br /><input type="number" defaultValue={secInRate}
```
Should this be in a label tag?

JustLe56 replied:
Fixed as of 8b4b241

huss reviewed (changes requested):
I need to play with this some more but I wanted to get out my first review comments now:

- There are a number of comments. I did some editing but left some to be done. I would like your thoughts and to know if you are able to do the changes (as appropriate/agree).
- It would be nice if the items in the create/edit modals spanned most of the horizontal area rather than a the limited amount they do now.
- Can the note area on the edit page be expandable as it is on the create page?
- We should show the name on the edit page. It is present on the create page.
- I think it would be fine for a non-admin to see the unit page but have no create or edit buttons. I tried to quickly change it so the option stayed on the main page but it did not let it go to the unit page. If you have time to fix this then great (or someone else needs to do it later).
- I edited some files that you should look at to see if you agree including
    - package-lock.json changed the version back to 1. This is fixed.
    - Not sure why the .vscode settings were changed (actaully empty was added) so put back.
    - Formatted some files.
    - Limited note to 30 characters on main unit page so not too big.
    - Modified some comments and spelling.
    - Fix a few variable names.
    - Modified unit-car-page.css so the card is smaller so more fit on a page. The project needs to be sure it is consistent across the new pages with cards.
    - Partly fixed up data.js for translations.

huss committed:
See PR for what was done

huss committed:
remove unique constaint

This was causing an issue when units were created/edited so it will
now be enforced in software. See the design document for more details.
The code still needs to add an update of cik/pik on unit changes.

huss committed:
remove left over trailing comma

huss commented:
Thanks to @andrew-cline for pointing out that there was an issue in editing units. It turned out to be related to a constraint on the units table:  
UNIQUE (type_of_unit, unit_index)  
See the design document for a discussion but the decision was made to remove this constraint. Basically, the unit_index is tricky to set and is invalid during the update of units until cik is updated. The code still needs to be altered to run the cik update but the constraint was removed in a recent commit to this PR. You can check if you have this constraint by doing:  
\d units  
and you will see something similar to  
"units_type_of_unit_unit_index_key" UNIQUE CONSTRAINT, btree (type_of_unit, unit_index)  
You can then remove this constraint with  
ALTER TABLE units DROP CONSTRAINT units_type_of_unit_unit_index_key;  
Once this is done you should no longer see an error about this. If you install the database from this commit forward then you should not have any issue.

JustLe56 committed:
updated translations

JustLe56 committed:
removed unit.create.new.unit

Commit:
updated create and edit unit forms to use enums

Commit:
removed unused import

Commit:
made changes based on comments on pull request

JustLe56 committed:
changed createUnitModalComponent form

JustLe56 committed:
Merge branch 'modalEdit_modalCreate' of https://github.com/andrew-cline/OED into modalEdit_modalCreate

huss committed:
updates (see details below)

- translated the labels of items
- translated enum menu items per upcoming changes in other parts
of the code base.
- translated Yes/No menu dropdown
- translated some buttons
- reorder edit page so items are same as create page and model

huss commented:
Many thanks to the team for addressing all the comments made. A second review indicates they are all done and it runs correctly. I have created the item "unit page updates" on the resource generalization project for what is left to do that is outside this work. This is ready to merge but I am waiting until review some other new modal pages to make sure no other changes are needed.

Note I made changes via a commit. The team is welcome (encouraged :-) to review them to make sure they all seem okay. The commit message lists what was done.

huss commented:
This PR is being held until we decide if any changes are needed to reconcile how other teams did similar work.

huss committed:
Merge branch 'resourceGeneralization' of https://github.com/OpenEnergyDashboard/OED into pr/andrew-cline/766-1

This should update the PR so it is now in sync with the
resourceGeneralization branch that is also in sync with development
branch. It should help those working on the modal issue.

huss commented:
React hooks have now been integrated into OED. The work in this PR has been integrated into that work and moved forward. Thus, the current plan is to close this PR and move forward with a new one.
